src/content/post/20161122_drupal_application_bottlenecks_avoid_with_profiling.md

```diff
@@ -0,0 +1,160 @@
+
+#### Application side
+
+On Drupal 7, we can use the [Performance Logging](https://www.drupal.org/project/performance) module to profile time spent to generate pages, memory consumption, and how many queries fire during page load. We only need to navigate our site to populate the data table. On Drupal 8 the great _Web Profiler_ module, recently merged into [Devel](https://www.drupal.org/project/devel) project is the best option.
+
+#### Environment side
+
+The environment (local but also production), with the right tools can become our better friend to help us to find reasons of application's bottlenecks. There is a lot of profiler, but an excellent choice is [Blackfire](http://blackfire.io/): free, flexible and very light (his agent can be installed on the production servers with a really really thin overhead). The feature which makes me love Blackfire are his profiler result: a simple and clear diagram of entire code execution flow from the request response to the final output: all function calls will be analyzed and counted to produce a detailed report that contain a lot of useful data like function's time execution, memory consumption, etc. Finally all this data will be collected and the results will be showed to us related to the global status: a function that takes three seconds to produce an output, can be a lower problem related to another function that takes only 0.5 second but it's called ninety time.
+Obviously, remember to puts on the mysql slow query log.
+
+#### An example:
+
+This is a Blackfire profiling output: [https://blackfire.io/profiles/f911c4ad-792b-4977-af51-d9e3b7649d24/graph](https://blackfire.io/profiles/f911c4ad-792b-4977-af51-d9e3b7649d24/graph)
+We can clearly see a 'red branch' that indicates us where the most of time is spent to generate the page and we can see that the _node_load_ function was called 200 times after the call to _dsc_module_block_view_ , uhm IMHO it's better take a look to this function:
+
+<pre>
+function dsc_module_block_view($delta = '') {
+  $block = array();
+  switch ($delta) {
+    case 'page_block':
+
+    $node_titles = [];
+    for ($i = 1; $i < 200; $i++) {
+      $node = node_load($i);
+      $node_titles[] = $node->title;
+    }
+    $content = '<ul>';
+    foreach ($node_titles as $node_title) {
+      $content .= '<li>;' . $node_title . '</li>';
+    }
+    $content .= '</ul>';
+    $block['subject'] = t('My block');
+    $block['content'] = array(
+      '#type' => 'markup',
+      '#markup' => $content,
+      '#title' => t('Last 100 content'),
+    );
+    break;
+  }
+  return $block;
+}
+</pre>
+
+We have a simple Drupal 7 application that display 200 node's titles, but in the wrong way.
```

Comment:
Questa informazione è quello che chiedevo sopra... la metterei prima in effetti!